{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: opencv-python in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from -r requirements.txt (line 1)) (4.7.0.68)\n",
      "Collecting face_recognition (from -r requirements.txt (line 2))\n",
      "  Using cached face_recognition-1.3.0-py2.py3-none-any.whl (15 kB)\n",
      "Requirement already satisfied: openai in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from -r requirements.txt (line 3)) (1.3.5)\n",
      "Collecting SpeechRecognition (from -r requirements.txt (line 4))\n",
      "  Using cached SpeechRecognition-3.10.0-py2.py3-none-any.whl (32.8 MB)\n",
      "Requirement already satisfied: firebase_admin in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from -r requirements.txt (line 5)) (6.2.0)\n",
      "Requirement already satisfied: numpy>=1.17.0 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from opencv-python->-r requirements.txt (line 1)) (1.24.1)\n",
      "Collecting face-recognition-models>=0.3.0 (from face_recognition->-r requirements.txt (line 2))\n",
      "  Using cached face_recognition_models-0.3.0-py2.py3-none-any.whl\n",
      "Collecting Click>=6.0 (from face_recognition->-r requirements.txt (line 2))\n",
      "  Using cached click-8.1.7-py3-none-any.whl.metadata (3.0 kB)\n",
      "Collecting dlib>=19.7 (from face_recognition->-r requirements.txt (line 2))\n",
      "  Using cached dlib-19.24.2.tar.gz (11.8 MB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Requirement already satisfied: Pillow in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from face_recognition->-r requirements.txt (line 2)) (9.4.0)\n",
      "Requirement already satisfied: anyio<4,>=3.5.0 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from openai->-r requirements.txt (line 3)) (3.7.1)\n",
      "Requirement already satisfied: distro<2,>=1.7.0 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from openai->-r requirements.txt (line 3)) (1.8.0)\n",
      "Requirement already satisfied: httpx<1,>=0.23.0 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from openai->-r requirements.txt (line 3)) (0.25.2)\n",
      "Requirement already satisfied: pydantic<3,>=1.9.0 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from openai->-r requirements.txt (line 3)) (2.5.2)\n",
      "Requirement already satisfied: tqdm>4 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from openai->-r requirements.txt (line 3)) (4.66.1)\n",
      "Requirement already satisfied: typing-extensions<5,>=4.5 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from openai->-r requirements.txt (line 3)) (4.8.0)\n",
      "Requirement already satisfied: requests>=2.26.0 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from SpeechRecognition->-r requirements.txt (line 4)) (2.28.1)\n",
      "Requirement already satisfied: cachecontrol>=0.12.6 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from firebase_admin->-r requirements.txt (line 5)) (0.13.1)\n",
      "Requirement already satisfied: google-api-python-client>=1.7.8 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from firebase_admin->-r requirements.txt (line 5)) (2.108.0)\n",
      "Requirement already satisfied: google-cloud-storage>=1.37.1 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from firebase_admin->-r requirements.txt (line 5)) (2.13.0)\n",
      "Requirement already satisfied: pyjwt>=2.5.0 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from pyjwt[crypto]>=2.5.0->firebase_admin->-r requirements.txt (line 5)) (2.8.0)\n",
      "Requirement already satisfied: google-api-core<3.0.0dev,>=1.22.1 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from google-api-core[grpc]<3.0.0dev,>=1.22.1; platform_python_implementation != \"PyPy\"->firebase_admin->-r requirements.txt (line 5)) (2.14.0)\n",
      "Requirement already satisfied: google-cloud-firestore>=2.9.1 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from firebase_admin->-r requirements.txt (line 5)) (2.13.1)\n",
      "Requirement already satisfied: idna>=2.8 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from anyio<4,>=3.5.0->openai->-r requirements.txt (line 3)) (3.4)\n",
      "Requirement already satisfied: sniffio>=1.1 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from anyio<4,>=3.5.0->openai->-r requirements.txt (line 3)) (1.3.0)\n",
      "Requirement already satisfied: exceptiongroup in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from anyio<4,>=3.5.0->openai->-r requirements.txt (line 3)) (1.2.0)\n",
      "Requirement already satisfied: msgpack>=0.5.2 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from cachecontrol>=0.12.6->firebase_admin->-r requirements.txt (line 5)) (1.0.7)\n",
      "Requirement already satisfied: colorama in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from Click>=6.0->face_recognition->-r requirements.txt (line 2)) (0.4.6)\n",
      "Requirement already satisfied: googleapis-common-protos<2.0.dev0,>=1.56.2 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from google-api-core<3.0.0dev,>=1.22.1->google-api-core[grpc]<3.0.0dev,>=1.22.1; platform_python_implementation != \"PyPy\"->firebase_admin->-r requirements.txt (line 5)) (1.61.0)\n",
      "Requirement already satisfied: protobuf!=3.20.0,!=3.20.1,!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0.dev0,>=3.19.5 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from google-api-core<3.0.0dev,>=1.22.1->google-api-core[grpc]<3.0.0dev,>=1.22.1; platform_python_implementation != \"PyPy\"->firebase_admin->-r requirements.txt (line 5)) (4.25.1)\n",
      "Requirement already satisfied: google-auth<3.0.dev0,>=2.14.1 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from google-api-core<3.0.0dev,>=1.22.1->google-api-core[grpc]<3.0.0dev,>=1.22.1; platform_python_implementation != \"PyPy\"->firebase_admin->-r requirements.txt (line 5)) (2.23.4)\n",
      "Requirement already satisfied: grpcio<2.0dev,>=1.33.2 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from google-api-core[grpc]<3.0.0dev,>=1.22.1; platform_python_implementation != \"PyPy\"->firebase_admin->-r requirements.txt (line 5)) (1.59.3)\n",
      "Requirement already satisfied: grpcio-status<2.0.dev0,>=1.33.2 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from google-api-core[grpc]<3.0.0dev,>=1.22.1; platform_python_implementation != \"PyPy\"->firebase_admin->-r requirements.txt (line 5)) (1.59.3)\n",
      "Requirement already satisfied: httplib2<1.dev0,>=0.15.0 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from google-api-python-client>=1.7.8->firebase_admin->-r requirements.txt (line 5)) (0.22.0)\n",
      "Requirement already satisfied: google-auth-httplib2>=0.1.0 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from google-api-python-client>=1.7.8->firebase_admin->-r requirements.txt (line 5)) (0.1.1)\n",
      "Requirement already satisfied: uritemplate<5,>=3.0.1 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from google-api-python-client>=1.7.8->firebase_admin->-r requirements.txt (line 5)) (4.1.1)\n",
      "Requirement already satisfied: google-cloud-core<3.0.0dev,>=1.4.1 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from google-cloud-firestore>=2.9.1->firebase_admin->-r requirements.txt (line 5)) (2.3.3)\n",
      "Requirement already satisfied: proto-plus<2.0.0dev,>=1.22.0 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from google-cloud-firestore>=2.9.1->firebase_admin->-r requirements.txt (line 5)) (1.22.3)\n",
      "Requirement already satisfied: google-resumable-media>=2.6.0 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from google-cloud-storage>=1.37.1->firebase_admin->-r requirements.txt (line 5)) (2.6.0)\n",
      "Requirement already satisfied: google-crc32c<2.0dev,>=1.0 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from google-cloud-storage>=1.37.1->firebase_admin->-r requirements.txt (line 5)) (1.5.0)\n",
      "Requirement already satisfied: certifi in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from httpx<1,>=0.23.0->openai->-r requirements.txt (line 3)) (2022.12.7)\n",
      "Requirement already satisfied: httpcore==1.* in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from httpx<1,>=0.23.0->openai->-r requirements.txt (line 3)) (1.0.2)\n",
      "Requirement already satisfied: h11<0.15,>=0.13 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from httpcore==1.*->httpx<1,>=0.23.0->openai->-r requirements.txt (line 3)) (0.14.0)\n",
      "Requirement already satisfied: annotated-types>=0.4.0 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from pydantic<3,>=1.9.0->openai->-r requirements.txt (line 3)) (0.6.0)\n",
      "Requirement already satisfied: pydantic-core==2.14.5 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from pydantic<3,>=1.9.0->openai->-r requirements.txt (line 3)) (2.14.5)\n",
      "Requirement already satisfied: cryptography>=3.4.0 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from pyjwt[crypto]>=2.5.0->firebase_admin->-r requirements.txt (line 5)) (41.0.5)\n",
      "Requirement already satisfied: charset-normalizer<3,>=2 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from requests>=2.26.0->SpeechRecognition->-r requirements.txt (line 4)) (2.1.1)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from requests>=2.26.0->SpeechRecognition->-r requirements.txt (line 4)) (1.26.14)\n",
      "Requirement already satisfied: cffi>=1.12 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from cryptography>=3.4.0->pyjwt[crypto]>=2.5.0->firebase_admin->-r requirements.txt (line 5)) (1.16.0)\n",
      "Requirement already satisfied: cachetools<6.0,>=2.0.0 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from google-auth<3.0.dev0,>=2.14.1->google-api-core<3.0.0dev,>=1.22.1->google-api-core[grpc]<3.0.0dev,>=1.22.1; platform_python_implementation != \"PyPy\"->firebase_admin->-r requirements.txt (line 5)) (5.3.0)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from google-auth<3.0.dev0,>=2.14.1->google-api-core<3.0.0dev,>=1.22.1->google-api-core[grpc]<3.0.0dev,>=1.22.1; platform_python_implementation != \"PyPy\"->firebase_admin->-r requirements.txt (line 5)) (0.2.8)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from google-auth<3.0.dev0,>=2.14.1->google-api-core<3.0.0dev,>=1.22.1->google-api-core[grpc]<3.0.0dev,>=1.22.1; platform_python_implementation != \"PyPy\"->firebase_admin->-r requirements.txt (line 5)) (4.9)\n",
      "Requirement already satisfied: pyparsing!=3.0.0,!=3.0.1,!=3.0.2,!=3.0.3,<4,>=2.4.2 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from httplib2<1.dev0,>=0.15.0->google-api-python-client>=1.7.8->firebase_admin->-r requirements.txt (line 5)) (3.0.9)\n",
      "Requirement already satisfied: pycparser in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from cffi>=1.12->cryptography>=3.4.0->pyjwt[crypto]>=2.5.0->firebase_admin->-r requirements.txt (line 5)) (2.21)\n",
      "Requirement already satisfied: pyasn1<0.5.0,>=0.4.6 in c:\\users\\james\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from pyasn1-modules>=0.2.1->google-auth<3.0.dev0,>=2.14.1->google-api-core<3.0.0dev,>=1.22.1->google-api-core[grpc]<3.0.0dev,>=1.22.1; platform_python_implementation != \"PyPy\"->firebase_admin->-r requirements.txt (line 5)) (0.4.8)\n",
      "Using cached click-8.1.7-py3-none-any.whl (97 kB)\n",
      "Building wheels for collected packages: dlib\n",
      "  Building wheel for dlib (pyproject.toml): started\n",
      "  Building wheel for dlib (pyproject.toml): finished with status 'error'\n",
      "Failed to build dlib\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  error: subprocess-exited-with-error\n",
      "  \n",
      "  × Building wheel for dlib (pyproject.toml) did not run successfully.\n",
      "  │ exit code: 1\n",
      "  ╰─> [73 lines of output]\n",
      "      running bdist_wheel\n",
      "      running build\n",
      "      running build_ext\n",
      "      <string>:125: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.\n",
      "      Building extension for Python 3.9.13 (tags/v3.9.13:6de2ca5, May 17 2022, 16:36:42) [MSC v.1929 64 bit (AMD64)]\n",
      "      Invoking CMake setup: 'cmake C:\\Users\\james\\AppData\\Local\\Temp\\pip-install-02hjb3q_\\dlib_b6f50943a1b24367ae1045d06dcb3a5c\\tools\\python -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=C:\\Users\\james\\AppData\\Local\\Temp\\pip-install-02hjb3q_\\dlib_b6f50943a1b24367ae1045d06dcb3a5c\\build\\lib.win-amd64-cpython-39 -DPYTHON_EXECUTABLE=C:\\Users\\james\\AppData\\Local\\Microsoft\\WindowsApps\\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\\python.exe -DCMAKE_LIBRARY_OUTPUT_DIRECTORY_RELEASE=C:\\Users\\james\\AppData\\Local\\Temp\\pip-install-02hjb3q_\\dlib_b6f50943a1b24367ae1045d06dcb3a5c\\build\\lib.win-amd64-cpython-39 -A x64'\n",
      "      -- Building for: NMake Makefiles\n",
      "      CMake Error at CMakeLists.txt:5 (message):\n",
      "      \n",
      "      \n",
      "      \n",
      "        !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "      \n",
      "      \n",
      "        You must use Visual Studio to build a python extension on windows.  If you\n",
      "        are getting this error it means you have not installed Visual C++.  Note\n",
      "        that there are many flavors of Visual Studio, like Visual Studio for C#\n",
      "        development.  You need to install Visual Studio for C++.\n",
      "      \n",
      "      \n",
      "        !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "      \n",
      "      \n",
      "      \n",
      "      \n",
      "      -- Configuring incomplete, errors occurred!\n",
      "      Traceback (most recent call last):\n",
      "        File \"C:\\Users\\james\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python39\\site-packages\\pip\\_vendor\\pyproject_hooks\\_in_process\\_in_process.py\", line 353, in <module>\n",
      "          main()\n",
      "        File \"C:\\Users\\james\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python39\\site-packages\\pip\\_vendor\\pyproject_hooks\\_in_process\\_in_process.py\", line 335, in main\n",
      "          json_out['return_val'] = hook(**hook_input['kwargs'])\n",
      "        File \"C:\\Users\\james\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python39\\site-packages\\pip\\_vendor\\pyproject_hooks\\_in_process\\_in_process.py\", line 251, in build_wheel\n",
      "          return _build_backend().build_wheel(wheel_directory, config_settings,\n",
      "        File \"C:\\Users\\james\\AppData\\Local\\Temp\\pip-build-env-maeu6a3x\\overlay\\Lib\\site-packages\\setuptools\\build_meta.py\", line 404, in build_wheel\n",
      "          return self._build_with_temp_dir(\n",
      "        File \"C:\\Users\\james\\AppData\\Local\\Temp\\pip-build-env-maeu6a3x\\overlay\\Lib\\site-packages\\setuptools\\build_meta.py\", line 389, in _build_with_temp_dir\n",
      "          self.run_setup()\n",
      "        File \"C:\\Users\\james\\AppData\\Local\\Temp\\pip-build-env-maeu6a3x\\overlay\\Lib\\site-packages\\setuptools\\build_meta.py\", line 311, in run_setup\n",
      "          exec(code, locals())\n",
      "        File \"<string>\", line 218, in <module>\n",
      "        File \"C:\\Users\\james\\AppData\\Local\\Temp\\pip-build-env-maeu6a3x\\overlay\\Lib\\site-packages\\setuptools\\__init__.py\", line 103, in setup\n",
      "          return distutils.core.setup(**attrs)\n",
      "        File \"C:\\Users\\james\\AppData\\Local\\Temp\\pip-build-env-maeu6a3x\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\core.py\", line 185, in setup\n",
      "          return run_commands(dist)\n",
      "        File \"C:\\Users\\james\\AppData\\Local\\Temp\\pip-build-env-maeu6a3x\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\core.py\", line 201, in run_commands\n",
      "          dist.run_commands()\n",
      "        File \"C:\\Users\\james\\AppData\\Local\\Temp\\pip-build-env-maeu6a3x\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\dist.py\", line 969, in run_commands\n",
      "          self.run_command(cmd)\n",
      "        File \"C:\\Users\\james\\AppData\\Local\\Temp\\pip-build-env-maeu6a3x\\overlay\\Lib\\site-packages\\setuptools\\dist.py\", line 963, in run_command\n",
      "          super().run_command(command)\n",
      "        File \"C:\\Users\\james\\AppData\\Local\\Temp\\pip-build-env-maeu6a3x\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\dist.py\", line 988, in run_command\n",
      "          cmd_obj.run()\n",
      "        File \"C:\\Users\\james\\AppData\\Local\\Temp\\pip-build-env-maeu6a3x\\overlay\\Lib\\site-packages\\wheel\\bdist_wheel.py\", line 369, in run\n",
      "          self.run_command(\"build\")\n",
      "        File \"C:\\Users\\james\\AppData\\Local\\Temp\\pip-build-env-maeu6a3x\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\cmd.py\", line 318, in run_command\n",
      "          self.distribution.run_command(command)\n",
      "        File \"C:\\Users\\james\\AppData\\Local\\Temp\\pip-build-env-maeu6a3x\\overlay\\Lib\\site-packages\\setuptools\\dist.py\", line 963, in run_command\n",
      "          super().run_command(command)\n",
      "        File \"C:\\Users\\james\\AppData\\Local\\Temp\\pip-build-env-maeu6a3x\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\dist.py\", line 988, in run_command\n",
      "          cmd_obj.run()\n",
      "        File \"C:\\Users\\james\\AppData\\Local\\Temp\\pip-build-env-maeu6a3x\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\command\\build.py\", line 131, in run\n",
      "          self.run_command(cmd_name)\n",
      "        File \"C:\\Users\\james\\AppData\\Local\\Temp\\pip-build-env-maeu6a3x\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\cmd.py\", line 318, in run_command\n",
      "          self.distribution.run_command(command)\n",
      "        File \"C:\\Users\\james\\AppData\\Local\\Temp\\pip-build-env-maeu6a3x\\overlay\\Lib\\site-packages\\setuptools\\dist.py\", line 963, in run_command\n",
      "          super().run_command(command)\n",
      "        File \"C:\\Users\\james\\AppData\\Local\\Temp\\pip-build-env-maeu6a3x\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\dist.py\", line 988, in run_command\n",
      "          cmd_obj.run()\n",
      "        File \"<string>\", line 130, in run\n",
      "        File \"<string>\", line 167, in build_extension\n",
      "        File \"C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.9_3.9.3568.0_x64__qbz5n2kfra8p0\\lib\\subprocess.py\", line 373, in check_call\n",
      "          raise CalledProcessError(retcode, cmd)\n",
      "      subprocess.CalledProcessError: Command '['cmake', 'C:\\\\Users\\\\james\\\\AppData\\\\Local\\\\Temp\\\\pip-install-02hjb3q_\\\\dlib_b6f50943a1b24367ae1045d06dcb3a5c\\\\tools\\\\python', '-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=C:\\\\Users\\\\james\\\\AppData\\\\Local\\\\Temp\\\\pip-install-02hjb3q_\\\\dlib_b6f50943a1b24367ae1045d06dcb3a5c\\\\build\\\\lib.win-amd64-cpython-39', '-DPYTHON_EXECUTABLE=C:\\\\Users\\\\james\\\\AppData\\\\Local\\\\Microsoft\\\\WindowsApps\\\\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\\\\python.exe', '-DCMAKE_LIBRARY_OUTPUT_DIRECTORY_RELEASE=C:\\\\Users\\\\james\\\\AppData\\\\Local\\\\Temp\\\\pip-install-02hjb3q_\\\\dlib_b6f50943a1b24367ae1045d06dcb3a5c\\\\build\\\\lib.win-amd64-cpython-39', '-A', 'x64']' returned non-zero exit status 1.\n",
      "      [end of output]\n",
      "  \n",
      "  note: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "  ERROR: Failed building wheel for dlib\n",
      "ERROR: Could not build wheels for dlib, which is required to install pyproject.toml-based projects\n"
     ]
    }
   ],
   "source": [
    "!pip3 install -r requirements.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'face_recognition'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\james\\OneDrive\\Desktop\\hackathon\\hackWestern2023\\hackwestern23\\test.ipynb Cell 2\u001b[0m line \u001b[0;36m2\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/james/OneDrive/Desktop/hackathon/hackWestern2023/hackwestern23/test.ipynb#W1sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mcv2\u001b[39;00m\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/james/OneDrive/Desktop/hackathon/hackWestern2023/hackwestern23/test.ipynb#W1sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mface_recognition\u001b[39;00m\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/james/OneDrive/Desktop/hackathon/hackWestern2023/hackwestern23/test.ipynb#W1sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mnumpy\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39mnp\u001b[39;00m\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/james/OneDrive/Desktop/hackathon/hackWestern2023/hackwestern23/test.ipynb#W1sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mspeech_recognition\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39msr\u001b[39;00m\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'face_recognition'"
     ]
    }
   ],
   "source": [
    "import cv2\n",
    "import face_recognition\n",
    "import numpy as np\n",
    "import speech_recognition as sr\n",
    "import threading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb Cell 3\u001b[0m line \u001b[0;36m8\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#W2sZmlsZQ%3D%3D?line=4'>5</a>\u001b[0m xander_encoding \u001b[39m=\u001b[39m face_recognition\u001b[39m.\u001b[39mface_encodings(xander_image)[\u001b[39m0\u001b[39m]\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#W2sZmlsZQ%3D%3D?line=6'>7</a>\u001b[0m james_image \u001b[39m=\u001b[39m face_recognition\u001b[39m.\u001b[39mload_image_file(\u001b[39m\"\u001b[39m\u001b[39mjames.jpg\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#W2sZmlsZQ%3D%3D?line=7'>8</a>\u001b[0m james_encoding \u001b[39m=\u001b[39m face_recognition\u001b[39m.\u001b[39;49mface_encodings(james_image)[\u001b[39m0\u001b[39;49m]\n",
      "\u001b[0;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "rachel_image = face_recognition.load_image_file(\"rachel.jpg\")\n",
    "rachel_encoding = face_recognition.face_encodings(rachel_image)[0]\n",
    "\n",
    "xander_image = face_recognition.load_image_file(\"xander.jpg\")\n",
    "xander_encoding = face_recognition.face_encodings(xander_image)[0]\n",
    "\n",
    "james_image = face_recognition.load_image_file(\"james.jpg\")\n",
    "james_encoding = face_recognition.face_encodings(james_image)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "known_face_encodings = [\n",
    "    rachel_encoding,\n",
    "    xander_encoding\n",
    "]\n",
    "known_face_names = [\n",
    "    \"Rachel Chen\",\n",
    "    \"Xander Chin\"\n",
    "]\n",
    "\n",
    "# Initialize some variables\n",
    "face_locations = []\n",
    "face_encodings = []\n",
    "face_names = []\n",
    "process_this_frame = True\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "New face detected, starting counter\n",
      "Counter for a face: 2\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "(-0.05152151361107826, 0.07755845040082932, 0.037228044122457504, -0.1537841409444809, -0.1140672042965889, -0.04546017944812775, -0.0634378045797348, -0.1530667543411255, 0.17138054966926575, -0.13944341242313385, 0.20998671650886536, -0.13667771220207214, -0.19002896547317505, -0.024973824620246887, -0.08587469160556793, 0.19186772406101227, -0.23813731968402863, -0.11467508226633072, 0.019088704138994217, 0.041823990643024445, 0.08768168836832047, -0.020468806847929955, 0.02065938152372837, 0.06548968702554703, -0.09012087434530258, -0.3156745433807373, -0.10985115170478821, -0.08997515588998795, -0.027399679645895958, -0.03621296212077141, -0.010294174775481224, 0.05813237652182579, -0.1944642961025238, -0.054869212210178375, 0.03172243759036064, 0.0672779530286789, -0.026663225144147873, -0.11980731040239334, 0.17128312587738037, -0.028815485537052155, -0.2411441206932068, 0.023737849667668343, 0.08677922934293747, 0.22546705603599548, 0.14355520904064178, 0.050557494163513184, 8.129249181365594e-05, -0.15787258744239807, 0.16162222623825073, -0.1649923473596573, 0.014311198145151138, 0.0870680883526802, 0.0800555869936943, 0.01054998766630888, 0.036994196474552155, -0.11460284888744354, 0.1109628826379776, 0.10743241757154465, -0.18027599155902863, -0.004275428131222725, 0.1548449695110321, -0.07686235755681992, 0.032953523099422455, -0.11653625220060349, 0.2782043516635895, 0.0658014565706253, -0.10622162371873856, -0.13686221837997437, 0.1297142505645752, -0.09354323148727417, -0.07566944509744644, 0.03218246251344681, -0.1360185444355011, -0.15835154056549072, -0.2654971778392792, -0.04022299870848656, 0.34390121698379517, 0.11891961842775345, -0.1344677060842514, 0.07517266273498535, -0.029279615730047226, 0.0012822533026337624, 0.09645490348339081, 0.17958679795265198, -0.01673092134296894, 0.05054197460412979, -0.06722623854875565, 0.005643901415169239, 0.21888235211372375, -0.09726598113775253, 0.01343133207410574, 0.2324959635734558, -0.04531138017773628, 0.031742341816425323, -0.01515588816255331, -0.048964131623506546, -0.12373560667037964, 0.08602281659841537, -0.11980096250772476, 0.022038554772734642, 0.034029919654130936, -0.010542774572968483, 0.002733161672949791, 0.14666740596294403, -0.20734567940235138, 0.08482904732227325, -0.05703805014491081, -0.012440210208296776, 0.005961295682936907, -0.04660738259553909, -0.05353345721960068, -0.04394499212503433, 0.1094503328204155, -0.23441798985004425, 0.15311893820762634, 0.11986995488405228, 0.08891163021326065, 0.10396375507116318, 0.08945430815219879, 0.059032876044511795, 0.006128387525677681, -0.009621730074286461, -0.24973727762699127, -0.012562062591314316, 0.13774782419204712, -0.04741066321730614, 0.11951059848070145, 0.004901561886072159)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb Cell 5\u001b[0m line \u001b[0;36m5\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#W3sZmlsZQ%3D%3D?line=54'>55</a>\u001b[0m     \u001b[39mprint\u001b[39m(\u001b[39m\"\u001b[39m\u001b[39mNew face detected, starting counter\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#W3sZmlsZQ%3D%3D?line=55'>56</a>\u001b[0m     unrecognized_faces[\u001b[39mtuple\u001b[39m(face_encoding)] \u001b[39m=\u001b[39m {\u001b[39m'\u001b[39m\u001b[39mcounter\u001b[39m\u001b[39m'\u001b[39m: \u001b[39m1\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mencoding\u001b[39m\u001b[39m'\u001b[39m: face_encoding}\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#W3sZmlsZQ%3D%3D?line=57'>58</a>\u001b[0m \u001b[39mif\u001b[39;00m unrecognized_faces[face_key][\u001b[39m'\u001b[39m\u001b[39mcounter\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m>\u001b[39m unrecognized_threshold:\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#W3sZmlsZQ%3D%3D?line=58'>59</a>\u001b[0m     name \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mUnnamed Person\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#W3sZmlsZQ%3D%3D?line=59'>60</a>\u001b[0m     face_encoding_list \u001b[39m=\u001b[39m unrecognized_faces[face_key][\u001b[39m'\u001b[39m\u001b[39mencoding\u001b[39m\u001b[39m'\u001b[39m]\u001b[39m.\u001b[39mtolist()\n",
      "\u001b[0;31mKeyError\u001b[0m: (-0.05152151361107826, 0.07755845040082932, 0.037228044122457504, -0.1537841409444809, -0.1140672042965889, -0.04546017944812775, -0.0634378045797348, -0.1530667543411255, 0.17138054966926575, -0.13944341242313385, 0.20998671650886536, -0.13667771220207214, -0.19002896547317505, -0.024973824620246887, -0.08587469160556793, 0.19186772406101227, -0.23813731968402863, -0.11467508226633072, 0.019088704138994217, 0.041823990643024445, 0.08768168836832047, -0.020468806847929955, 0.02065938152372837, 0.06548968702554703, -0.09012087434530258, -0.3156745433807373, -0.10985115170478821, -0.08997515588998795, -0.027399679645895958, -0.03621296212077141, -0.010294174775481224, 0.05813237652182579, -0.1944642961025238, -0.054869212210178375, 0.03172243759036064, 0.0672779530286789, -0.026663225144147873, -0.11980731040239334, 0.17128312587738037, -0.028815485537052155, -0.2411441206932068, 0.023737849667668343, 0.08677922934293747, 0.22546705603599548, 0.14355520904064178, 0.050557494163513184, 8.129249181365594e-05, -0.15787258744239807, 0.16162222623825073, -0.1649923473596573, 0.014311198145151138, 0.0870680883526802, 0.0800555869936943, 0.01054998766630888, 0.036994196474552155, -0.11460284888744354, 0.1109628826379776, 0.10743241757154465, -0.18027599155902863, -0.004275428131222725, 0.1548449695110321, -0.07686235755681992, 0.032953523099422455, -0.11653625220060349, 0.2782043516635895, 0.0658014565706253, -0.10622162371873856, -0.13686221837997437, 0.1297142505645752, -0.09354323148727417, -0.07566944509744644, 0.03218246251344681, -0.1360185444355011, -0.15835154056549072, -0.2654971778392792, -0.04022299870848656, 0.34390121698379517, 0.11891961842775345, -0.1344677060842514, 0.07517266273498535, -0.029279615730047226, 0.0012822533026337624, 0.09645490348339081, 0.17958679795265198, -0.01673092134296894, 0.05054197460412979, -0.06722623854875565, 0.005643901415169239, 0.21888235211372375, -0.09726598113775253, 0.01343133207410574, 0.2324959635734558, -0.04531138017773628, 0.031742341816425323, -0.01515588816255331, -0.048964131623506546, -0.12373560667037964, 0.08602281659841537, -0.11980096250772476, 0.022038554772734642, 0.034029919654130936, -0.010542774572968483, 0.002733161672949791, 0.14666740596294403, -0.20734567940235138, 0.08482904732227325, -0.05703805014491081, -0.012440210208296776, 0.005961295682936907, -0.04660738259553909, -0.05353345721960068, -0.04394499212503433, 0.1094503328204155, -0.23441798985004425, 0.15311893820762634, 0.11986995488405228, 0.08891163021326065, 0.10396375507116318, 0.08945430815219879, 0.059032876044511795, 0.006128387525677681, -0.009621730074286461, -0.24973727762699127, -0.012562062591314316, 0.13774782419204712, -0.04741066321730614, 0.11951059848070145, 0.004901561886072159)"
     ]
    }
   ],
   "source": [
    "def fetch_encodings_from_firestore():\n",
    "    fetched_encodings = []\n",
    "    fetched_names = []\n",
    "    \n",
    "    # Fetch data from Firestore\n",
    "    docs = db.collection('people').stream()\n",
    "    for doc in docs:\n",
    "        data = doc.to_dict()\n",
    "        if 'image_enc' in data and 'name' in data:\n",
    "            fetched_encodings.append(np.array(data['image_enc']))\n",
    "            fetched_names.append(data['name'])\n",
    "    \n",
    "    return fetched_encodings, fetched_names\n",
    "\n",
    "def find_similar_face_key(face_encoding, faces_dict, tolerance=0.6):\n",
    "    for face_key in faces_dict.keys():\n",
    "        if np.linalg.norm(np.array(face_key) - face_encoding) < tolerance:\n",
    "            return face_key\n",
    "    return None\n",
    "\n",
    "\n",
    "face_encodings_from_db, face_names_from_db = fetch_encodings_from_firestore()\n",
    "\n",
    "# Dictionary to track unrecognized faces\n",
    "unrecognized_faces = {}\n",
    "unrecognized_threshold = 2  # Number of frames to confirm an unrecognized face\n",
    "\n",
    "video_capture = cv2.VideoCapture(1)\n",
    "\n",
    "while True:\n",
    "    ret, frame = video_capture.read()\n",
    "    small_frame = cv2.resize(frame, (0, 0), fx=0.25, fy=0.25)\n",
    "    rgb_small_frame = np.ascontiguousarray(small_frame[:, :, ::-1])\n",
    "\n",
    "    face_locations = face_recognition.face_locations(rgb_small_frame)\n",
    "    face_encodings = face_recognition.face_encodings(rgb_small_frame, face_locations)\n",
    "\n",
    "    face_names = []\n",
    "    for face_encoding in face_encodings:\n",
    "        matches = face_recognition.compare_faces(face_encodings_from_db, face_encoding, tolerance=0.5)\n",
    "\n",
    "        if True in matches:\n",
    "            first_match_index = matches.index(True)\n",
    "            name = face_names_from_db[first_match_index]\n",
    "            similar_key = find_similar_face_key(face_encoding, unrecognized_faces)\n",
    "            if similar_key is not None:\n",
    "                del unrecognized_faces[similar_key]  # Remove from unrecognized as it's now recognized\n",
    "        else:\n",
    "            similar_key = find_similar_face_key(face_encoding, unrecognized_faces)\n",
    "\n",
    "            if similar_key is not None:\n",
    "                unrecognized_faces[similar_key]['counter'] += 1\n",
    "                print(f\"Counter for a face: {unrecognized_faces[similar_key]['counter']}\")\n",
    "            else:\n",
    "                print(\"New face detected, starting counter\")\n",
    "                unrecognized_faces[tuple(face_encoding)] = {'counter': 1, 'encoding': face_encoding}\n",
    "\n",
    "            if similar_key is not None and unrecognized_faces[similar_key]['counter'] > unrecognized_threshold:\n",
    "                name = \"Unnamed Person\"\n",
    "                face_encoding_list = unrecognized_faces[similar_key]['encoding'].tolist()\n",
    "                doc_ref = db.collection('people').document()\n",
    "                doc_ref.set({\n",
    "                    'image_enc': face_encoding_list,\n",
    "                    'name': name,\n",
    "                })\n",
    "                face_encodings_from_db, face_names_from_db = fetch_encodings_from_firestore()\n",
    "                del unrecognized_faces[similar_key]\n",
    "            else:\n",
    "                name = \"Processing...\"\n",
    "\n",
    "        face_names.append(name)\n",
    "\n",
    "    for (top, right, bottom, left), name in zip(face_locations, face_names):\n",
    "        top *= 4\n",
    "        right *= 4\n",
    "        bottom *= 4\n",
    "        left *= 4\n",
    "        cv2.rectangle(frame, (left, top), (right, bottom), (0, 0, 255), 2)\n",
    "        cv2.rectangle(frame, (left, bottom - 35), (right, bottom), (0, 0, 255), cv2.FILLED)\n",
    "        font = cv2.FONT_HERSHEY_DUPLEX\n",
    "        cv2.putText(frame, name, (left + 6, bottom - 6), font, 1.0, (255, 255, 255), 1)\n",
    "\n",
    "    cv2.imshow('Video', frame)\n",
    "    if cv2.waitKey(1) & 0xFF == ord('q'):\n",
    "        break\n",
    "\n",
    "video_capture.release()\n",
    "cv2.destroyAllWindows()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def audio_transcription():\n",
    "    recognizer = sr.Recognizer()\n",
    "    with sr.Microphone() as source:\n",
    "        print(\"Say something!\")\n",
    "        while True:\n",
    "            audio = recognizer.listen(source)\n",
    "            try:\n",
    "                # Transcribe the audio to text\n",
    "                text = recognizer.recognize_google(audio)\n",
    "                print(\"Transcription: \" + text)\n",
    "            except sr.UnknownValueError:\n",
    "                print(\"Could not understand audio\")\n",
    "            except sr.RequestError as e:\n",
    "                print(\"Error; {0}\".format(e))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "Could not find PyAudio; check installation",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "File \u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/speech_recognition/__init__.py:108\u001b[0m, in \u001b[0;36mMicrophone.get_pyaudio\u001b[0;34m()\u001b[0m\n\u001b[1;32m    107\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 108\u001b[0m     \u001b[39mimport\u001b[39;00m \u001b[39mpyaudio\u001b[39;00m\n\u001b[1;32m    109\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mImportError\u001b[39;00m:\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'pyaudio'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[1;32m/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb Cell 7\u001b[0m line \u001b[0;36m6\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#X12sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m camera_thread\u001b[39m.\u001b[39mstart()\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#X12sZmlsZQ%3D%3D?line=4'>5</a>\u001b[0m \u001b[39m# Running audio transcription in the main thread\u001b[39;00m\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#X12sZmlsZQ%3D%3D?line=5'>6</a>\u001b[0m audio_transcription()\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#X12sZmlsZQ%3D%3D?line=7'>8</a>\u001b[0m \u001b[39m# Wait for the camera thread to finish\u001b[39;00m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#X12sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m camera_thread\u001b[39m.\u001b[39mjoin()\n",
      "\u001b[1;32m/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb Cell 7\u001b[0m line \u001b[0;36m3\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#X12sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39maudio_transcription\u001b[39m():\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#X12sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m     recognizer \u001b[39m=\u001b[39m sr\u001b[39m.\u001b[39mRecognizer()\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#X12sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m     \u001b[39mwith\u001b[39;00m sr\u001b[39m.\u001b[39;49mMicrophone() \u001b[39mas\u001b[39;00m source:\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#X12sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m         \u001b[39mprint\u001b[39m(\u001b[39m\"\u001b[39m\u001b[39mSay something!\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#X12sZmlsZQ%3D%3D?line=4'>5</a>\u001b[0m         \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n",
      "File \u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/speech_recognition/__init__.py:80\u001b[0m, in \u001b[0;36mMicrophone.__init__\u001b[0;34m(self, device_index, sample_rate, chunk_size)\u001b[0m\n\u001b[1;32m     77\u001b[0m \u001b[39massert\u001b[39;00m \u001b[39misinstance\u001b[39m(chunk_size, \u001b[39mint\u001b[39m) \u001b[39mand\u001b[39;00m chunk_size \u001b[39m>\u001b[39m \u001b[39m0\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mChunk size must be a positive integer\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m     79\u001b[0m \u001b[39m# set up PyAudio\u001b[39;00m\n\u001b[0;32m---> 80\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mpyaudio_module \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mget_pyaudio()\n\u001b[1;32m     81\u001b[0m audio \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mpyaudio_module\u001b[39m.\u001b[39mPyAudio()\n\u001b[1;32m     82\u001b[0m \u001b[39mtry\u001b[39;00m:\n",
      "File \u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/speech_recognition/__init__.py:110\u001b[0m, in \u001b[0;36mMicrophone.get_pyaudio\u001b[0;34m()\u001b[0m\n\u001b[1;32m    108\u001b[0m     \u001b[39mimport\u001b[39;00m \u001b[39mpyaudio\u001b[39;00m\n\u001b[1;32m    109\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mImportError\u001b[39;00m:\n\u001b[0;32m--> 110\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mAttributeError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mCould not find PyAudio; check installation\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    111\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mdistutils\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mversion\u001b[39;00m \u001b[39mimport\u001b[39;00m LooseVersion\n\u001b[1;32m    112\u001b[0m \u001b[39mif\u001b[39;00m LooseVersion(pyaudio\u001b[39m.\u001b[39m__version__) \u001b[39m<\u001b[39m LooseVersion(\u001b[39m\"\u001b[39m\u001b[39m0.2.11\u001b[39m\u001b[39m\"\u001b[39m):\n",
      "\u001b[0;31mAttributeError\u001b[0m: Could not find PyAudio; check installation"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread Thread-4:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/threading.py\", line 973, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/threading.py\", line 910, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/var/folders/q4/w5wd298x2j51mk7bkv38kmt80000gn/T/ipykernel_11648/1519530862.py\", line 55, in camera_operations\n",
      "cv2.error: Unknown C++ exception from OpenCV code\n"
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the the current cell or a previous cell. Please review the code in the cell(s) to identify a possible cause of the failure. Click <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. View Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "# Running camera operations in a separate thread\n",
    "camera_thread = threading.Thread(target=camera_operations)\n",
    "camera_thread.start()\n",
    "\n",
    "# Running audio transcription in the main thread\n",
    "audio_transcription()\n",
    "\n",
    "# Wait for the camera thread to finish\n",
    "camera_thread.join()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "OpenAIError",
     "evalue": "The api_key client option must be set either by passing api_key to the client or by setting the OPENAI_API_KEY environment variable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOpenAIError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[1;32m/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb Cell 8\u001b[0m line \u001b[0;36m2\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#W5sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mopenai\u001b[39;00m \u001b[39mimport\u001b[39;00m OpenAI\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#W5sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m client \u001b[39m=\u001b[39m OpenAI()\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#W5sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m openai\u001b[39m.\u001b[39mapi_key \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39msk-GBcT9AD2B4YEzw1se5qrT3BlbkFJzBH7RQACIjcepuVXMUBa\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/rachelchen/Documents/GitHub/hackwestern23/test.ipynb#W5sZmlsZQ%3D%3D?line=5'>6</a>\u001b[0m client \u001b[39m=\u001b[39m openai\u001b[39m.\u001b[39mOpenAI\u001b[39m.\u001b[39mclient()\n",
      "File \u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/openai/_client.py:93\u001b[0m, in \u001b[0;36mOpenAI.__init__\u001b[0;34m(self, api_key, organization, base_url, timeout, max_retries, default_headers, default_query, http_client, _strict_response_validation)\u001b[0m\n\u001b[1;32m     91\u001b[0m     api_key \u001b[39m=\u001b[39m os\u001b[39m.\u001b[39menviron\u001b[39m.\u001b[39mget(\u001b[39m\"\u001b[39m\u001b[39mOPENAI_API_KEY\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m     92\u001b[0m \u001b[39mif\u001b[39;00m api_key \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m---> 93\u001b[0m     \u001b[39mraise\u001b[39;00m OpenAIError(\n\u001b[1;32m     94\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mThe api_key client option must be set either by passing api_key to the client or by setting the OPENAI_API_KEY environment variable\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m     95\u001b[0m     )\n\u001b[1;32m     96\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mapi_key \u001b[39m=\u001b[39m api_key\n\u001b[1;32m     98\u001b[0m \u001b[39mif\u001b[39;00m organization \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "\u001b[0;31mOpenAIError\u001b[0m: The api_key client option must be set either by passing api_key to the client or by setting the OPENAI_API_KEY environment variable"
     ]
    }
   ],
   "source": [
    "from openai import OpenAI\n",
    "client = OpenAI()\n",
    "\n",
    "openai.api_key = \"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\"\n",
    "\n",
    "client = openai.OpenAI.client()\n",
    "\n",
    "audio_file = open(\"speech.mp3\", \"rb\")\n",
    "transcript = client.audio.transcriptions.create(\n",
    "  model=\"whisper-1\", \n",
    "  file=audio_file\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/rachelchen/Library/Python/3.9/lib/python/site-packages/urllib3/__init__.py:34: NotOpenSSLWarning: urllib3 v2 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<firebase_admin.App at 0x1102724f0>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import firebase_admin\n",
    "\n",
    "cred = firebase_admin.credentials.Certificate('hw23-e0512-firebase-adminsdk-3ax9k-293086f6f4.json')\n",
    "firebase_admin.initialize_app(cred)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from firebase_admin import firestore\n",
    "\n",
    "# Get a reference to the Firestore service\n",
    "db = firestore.client()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
